### PR TITLE
Fix Blackhole hang on RM reshape to [N,1]

### LINKED ISSUE #50191: [Bug] Blackhole: ttnn.reshape of a ROW_MAJOR tensor to [N, 1] hard-deadlocks the device
*Suggested labels: `blackhole`, `bug`, `data_movement`*

## Summary

On Blackhole, reshaping a large **ROW_MAJOR** `bfloat16` tensor to a column shape `[N, 1]`
(`N ≈ 250M`) on **interleaved DRAM** intermittently **hard-deadlocks the device**:
`ttnn.synchronize_device` never returns, **no exception is raised**, and only `SIGKILL` / `tt-smi -r`
recovers. It is **not** OOM — a genuine over-allocation raises a clean `TT_FATAL: Out of Memory`.

Root cause: reshaping to `[..., 1]` gives `dest_page_size_bytes = 1 * sizeof(bf16) = 2` bytes, far below
the Blackhole DRAM alignment of **64 bytes**. The RM reshape kernel then drains each source page into
2-byte destination pages, executing a `noc_async_write_barrier()` on **every** tiny write (~250M
barriered 2-byte interleaved-DRAM writes); a write completion eventually never returns and the barrier
waits forever.

## Environment

- Reproduced on **v0.70.0rc6**; the code path is **unchanged on `main` (v0.75.0-dev)**.
- Blackhole, single chip (`MeshDevice`), **interleaved DRAM**. Firmware 19.7.0, KMD 2.7.0.

## Reproduce

```python
import torch, ttnn
dev = ttnn.open_device(device_id=0)
T, G1 = 2500, 100001                       # rank [2500, 100001] bf16 TILE; N = T*G1 ≈ 250M
rank = ttnn.from_torch(torch.zeros(T, G1, dtype=torch.bfloat16),
                       dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=dev)
for i in range(60):
    tl = ttnn.to_layout(rank, ttnn.ROW_MAJOR_LAYOUT)
    rf = ttnn.reshape(tl, (T * G1, 1))     # [N,1] RM: dest_page_size = 2 bytes
    ttnn.synchronize_device(dev)           # <-- HANGS here at a nondeterministic iter
    ttnn.deallocate(rf); ttnn.deallocate(tl)
```

Reshaping to `[1, N]` instead never hangs. Intermittent — hangs at a **nondeterministic iteration**
(observed 8 / 14 / 24 / 43); `leak` / `scatter` / a sync between the two ops are not required — only the
`[N, 1]` (unit-last-dim) reshape is.

## Root cause

`ttnn/cpp/ttnn/operations/data_movement/reshape_view/device/device/rm_reshape_interleaved.cpp`, the
non-`can_be_clean` path (taken because `dest_page_size_bytes % 16 != 0`):

```cpp
uint32_t writable = dest_page_size_bytes - write_start_offset;   // = 2 for [N,1]
while (readable > 0) {
    noc_async_write_barrier();     // line 112: executed once per <=2-byte write, ~250M total
    ...
}
```

`reshape_rm_program_factory.cpp:45` computes `dest_page_size_bytes = output_log_shape[-1] * data_size`
with **no guard** for `dest_page_size_bytes < DRAM_ALIGNMENT` (64 B on Blackhole). The in-loop barrier
was introduced in #15572 (2024-12) and is unchanged since.

## Notes

No existing issue matches this path (surveyed); the closest, #40612, is a different
Wormhole / TILE-layout root cause. I've root-caused this and am happy to work on a fix/PR.

### PR DESCRIPTION
**Pull request by VirdhatchaniKN**
### Summary
* Fixes hard deadlock on Blackhole when reshaping ROW_MAJOR `bfloat16` tensors to unit last-dim shapes like `[N, 1]` on interleaved DRAM (`ttnn.synchronize_device` never returns; only `SIGKILL` / `tt-smi -r` recovers).
* Root cause: `dest_page_size_bytes = 2` takes the non-clean path and dual-issues NOC0+NOC1 with a barrier on every tiny DRAM write, triggering the Blackhole DRAM-arbiter hang (SYS-1419).
* Fix (factory + kernel, data_movement + Quasar):
  1. **Disable dual-kernel** when destination page is not NoC-aligned on DRAM (source alignment does not affect DRAM write safety).
  2. **Pipeline non-clean dest writes** via an L1 staging ring (barrier per batch, not per page); slot count is capped by available L1 (`lowest_occupied_compute_l1_address()`), falling back toward 1 slot for wide odd dests.
  3. **Pad** full-page non-clean writes to NOC write alignment (16B); safe because each interleaved page owns a private aligned slot.
  4. **Stage with read-datamover** `tt_memmove` so local staging does not drain in-flight DRAM dest writes (keeps ring batching effective).
  5. Quasar kernel keeps **DataflowBuffer** reserve/push APIs; slot ring offsets remain manual.

Fixes #50191
Verified the hanging test (Test file : [test_reshape_hang.py](https://github.com/user-attachments/files/30342097/test_reshape_hang.py) , Result : [result.txt](https://github.com/user-attachments/files/30342113/result.txt) ) 

### Pipeline status

- [x] Sanity
- [x] (Runtime) Sanity Test
- [x] BHPC
- [x] Nightly
- [x] All model test
- [x] Single card
- [x] T3K
- [x] Galaxy

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:virdhatchani/reshape_bug)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:virdhatchani/reshape_bug)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-sanity-tests.yaml?query=branch:virdhatchani/reshape_bug)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:virdhatchani/reshape_bug)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:virdhatchani/reshape_bug)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:virdhatchani/reshape_bug)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:virdhatchani/reshape_bug)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:virdhatchani/reshape_bug)

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/data_movement/reshape_view/device/reshape_rm_program_factory.cpp [hunk 1]
```diff
@@ -15,6 +19,42 @@ namespace ttnn::prim {
...
+constexpr uint32_t kSmallDestWriteSlots = 8;
+
+// Non-clean dest staging uses a multi-slot L1 ring. Cap slots by the per-core L1
+// budget so wide odd destinations (e.g. bf16 width 100001 from #50191) still fit.
+uint32_t choose_num_dest_write_slots(
+    IDevice* device,
+    bool pages_16b_aligned,
+    bool can_use_dual_kernel,
+    uint32_t cb_size0,
+    uint32_t dest_slot_size_bytes) {
```
[the remark this answers is not in the document]
Comment (VirdhatchaniKN) [reply ttnn/cpp/ttnn/operations/data_movement/reshape_view/device/reshape_rm_program_factory.cpp:42]: Skipping a shared header here. Quasar’s factory intentionally mirrors the data_movement one; extracting choose_num_dest_write_slots would only dedupe ~30 lines and add cross-op / CMake coupling. Left the helper local in both factories with a short comment.

**Review by Aswinmcw — APPROVED** [review]
Verified against #50191 — the fix targets the root cause directly: the per-page `noc_async_write_barrier()` is now gated behind `can_be_clean`, and the non-clean DRAM path batches the barrier over an L1 slot ring, disables dual-kernel on non-16B-aligned DRAM dests (SYS-1419), and pads writes to NOC alignment. Terminal drain barrier is in place and the last page barrier+returns in both the exact-fill and partial branches. Regression + accumulate/drain + L1-budget-shrink + Quasar cases all assert bit-exact. The current head already resolves the earlier bot flags (helper inlined, `use_read_datamover=true`, slot count L1-capped). LGTM.

**Comment by amorrisonTT** [comment]
Is there a link to SYS-1419, mentioned in the description? I want to understand the root cause.

Otherwise, LGTM

**Comment by VirdhatchaniKN** [comment]
> Is there a link to SYS-1419, mentioned in the description? I want to understand the root cause.

SYS-1419 is an internal Blackhole DRAM-arbiter hang. The in-repo description and disabled reproduction test are [here](https://github.com/tenstorrent/tt-metal/blob/02246470f9966fb091add6a76b84591551794158/tests/tt_metal/tt_metal/api/test_dram.cpp#L360-L366). When both NOCs send traffic to the same DRAM endpoint at a high rate, the arbiter may drop or stall a request, causing a barrier to wait indefinitely. This PR avoids that pattern by disabling the dual-kernel path for non-16-byte-aligned interleaved DRAM destination pages and by batching and padding the remaining small writes.

**Review by amorrisonTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/data_movement/reshape_view/device/device/rm_reshape_interleaved.cpp [hunk 2]
```diff
@@ -5,33 +5,47 @@
...
+6. write_start_offset
+7. nop
 */
 #include <stdint.h>
 #include "api/dataflow/dataflow_api.h"
 #include "api/dataflow/noc.h"
 #include "api/debug/dprint.h"  // required in all kernels using DPRINT
 #include "ttnn/operations/data_movement/common/kernels/common.hpp"
 
+FORCE_INLINE void acquire_dest_slot(Noc noc, uint32_t& slots_in_flight, const uint32_t num_dest_write_slots) {
```
Comment (fplavecTT) [inline ttnn/cpp/ttnn/operations/data_movement/reshape_view/device/device/rm_reshape_interleaved.cpp:39]: Both of these functions could use brief descriptions.

Same applies to equivalent functions in the `experimental` folder

File: ttnn/cpp/ttnn/operations/data_movement/reshape_view/device/reshape_rm_program_factory.cpp [hunk 3]
```diff
@@ -15,6 +19,46 @@ namespace ttnn::prim {
 
 using namespace tt::tt_metal;
 
+namespace {
+constexpr uint32_t kSmallDestWriteSlots = 8;
```
Comment (fplavecTT) [inline ttnn/cpp/ttnn/operations/data_movement/reshape_view/device/reshape_rm_program_factory.cpp]: Could use a comment why this is set to 8.
Also, all variables in this file use snake case, so it would be better to follow the convention: `small_dest_write_slots`

Same applies to equivalent variable in the `experimental` folder

File: ttnn/cpp/ttnn/operations/data_movement/reshape_view/device/reshape_rm_program_factory.cpp [hunk 4]
```diff
@@ -56,10 +100,23 @@ ProgramDescriptor ReshapeViewRMProgramFactory::create_descriptor(
         responsibility++;
     }
     const uint32_t cb_size0 = source_read_size_bytes;
-    const uint32_t cb_size1 = ((dest_page_size_bytes - 1) & MASK_64) + 80;
+    const uint32_t dest_slot_size_bytes = ((dest_page_size_bytes - 1) & MASK_64) + 80;
 
-    bool can_use_dual_kernel =
+    const bool pages_16b_aligned = (source_page_size_bytes % 16 == 0) && (dest_page_size_bytes % 16 == 0);
```
Comment (fplavecTT) [inline ttnn/cpp/ttnn/operations/data_movement/reshape_view/device/reshape_rm_program_factory.cpp]: Please add a comment or a meaningful identifier for the magic number 16.
If both `16`s are the same quantity, then it should definitely be an identifier so the two can't accidentally go out of sync.
Similarly, name of the variable `pages_16b_aligned` seems overly specific and subject to future churn if that value ever needs to change (same applies above in argument name to choose_num_dest_write_slots).

Same applies to  equivalent code in the `experimental` folder.

**Review by fplavecTT — APPROVED** [review]
Minor cleanups to comments/variables needed, otherwise good to go

**Review by bbradelTT — APPROVED** [review]

**Review by mateusznowakTT — APPROVED** [review]

**Comment by VirdhatchaniKN** [comment]
https://github.com/tenstorrent/tt-metal/actions/runs/31458530053 - passed
